### ISSUE TITLE
Circular Reference Detection does not respect XmlIgnoreAttribute

### ISSUE THREAD

**Issue by ddrdoubled** [wojtpl2/ExtendedXmlSerializer#190]
Seems when detecting Circular References [XmlgnoreAttribute] is not respected.

**Comment by Mike-EEE** [comment]
Boh... got a nifty test to show this off?  Are you saying that a field with `XmlgnoreAttribute` applied to it is still getting emitted?

**Comment by ddrdoubled** [comment]
Hi sorry, it is actually a different issue, please try to serialize TestClass, it seems reference equality within same level throwns Circular Reference Exception:
public class TestClass
{
		public TestClass()
		{
			Id = 1;
            var test = new TestClass2();
            Test1 = test;
            Test2 = test;
		}

		public int Id { get; set; }

        public TestClass2 Test1 { get; set; }

        public TestClass2 Test2 { get; set; }
	}

public class TestClass2
    {
        public int Id { get; set; }
    }

**Comment by Mike-EEE** [comment]
Ah, correct.  You will need to [enable references](https://github.com/wojtpl2/ExtendedXmlSerializer#object-reference-and-circular-reference) if they exist in your object graph. :)

**Comment by ddrdoubled** [comment]
ok, but that means i have to add a lot of configuration to my serialization. XmlSerializer works pretty smoothly with this, is it difficult to change default behaviour?

**Comment by Mike-EEE** [comment]
Ah sorry for the confusion @ddrdoubled ... not much configuration is necessary if you are simply wanting to enable references.  [Here is a file](https://github.com/wojtpl2/ExtendedXmlSerializer/blob/7aba7571c10b998b51e8bf879fa632746e6abb82/test/ExtendedXmlSerializer.Tests/ReportedIssues/Issue190Tests.cs) that shows you what is necessary, using your objects provided above.

Please let me know if you have any further questions/concerns and I will do my best to help you out.

**Comment by ddrdoubled** [comment]
Test case still throws Exception, this time "null reference exception no constructor" for Class2. Did your test file run through?

**Comment by Mike-EEE** [comment]
Oh, are you saying the tests that I provided are throwing an error?  They [passed on my machine](https://desu-usergeneratedcontent.xyz/g/image/1328/50/1328508781464.png) and [through CI](https://ci.appveyor.com/project/wojtpl2/extendedxmlserializer/build/258/tests), so they *seem* to work. 😂

If you are using another test with instructions that differ from the ones provided, please feel free to share and we'll see what's wrong.

**Comment by ddrdoubled** [comment]
That's strange, sorry to bother you: i am using this test:
var dataContract = new TestClass();
var xmlSerializer = new ConfigurationContainer().EnableReferences().Create();
var fileName = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments) + "//Test.xml";
FileStream file = File.Create(fileName);
xmlSerializer.Serialize(file, swapTradeDataContract);
file.Close();

TestClass and TestClass2 are the same, is there an issue with the extension method?

**Comment by ddrdoubled** [comment]
It is the EnableReferences() which throws already the exception

**Comment by ddrdoubled** [comment]
Entry Point is here:
public static IRootContext EnableReferences(this IRootContext @this)
		{
			@this.EnableRootInstances().With<ReferencesExtension>();
			return @this;
		}

**Comment by ddrdoubled** [comment]
Null Reference Exception is thrown here ( constructor is null) in class DefaultActivators

NewExpression Reference(Type parameter, TypeInfo typeInfo)
		{
			var constructor = _locator.Get(typeInfo);
			var parameters = constructor.GetParameters();
			var result = parameters.Length > 0
				? Expression.New(constructor, parameters.Select(Selector))
				: Expression.New(parameter);
			return result;
		}

**Comment by Mike-EEE** [comment]
Are you saying the exception is being thrown at `EnableReferences`?  From eyeballing it, you are serializing `swapTradeDataContract` which is not defined in the presented code, not that I can see.  What happens when you change that to `dataContact` instance?

It might be best to provide a PR with the failing test and I can take a look at it from there. 👍

**Comment by ddrdoubled** [comment]
yes the exception is already thrown by EnableReferences(). Sorry for the swapTradeDataContract i have corrected it.

**Comment by Mike-EEE** [comment]
Well, that is interesting.  Are you able to run the tests from the repo here alright?  If you can submit a PR with the failing test I will take a look into it.
